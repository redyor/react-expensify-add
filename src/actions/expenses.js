import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';
// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispach, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then((ref) => {
        dispach(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
        // console('Saving data now dispatch?');
      })
      .catch((e) => {
        //console.log('Error', e);
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

// Remove expense from firebase
export const startRemoveExpense = ({ id }) => {
  return (dispach, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispach(
          removeExpense({
            id,
          })
        );
        //  console('Remove data from database.');
      })
      .catch((e) => {
        // console.log('Error', e);
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
  // console.log(id, description, note, amount, createdAt);
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        // console.log('Data Changed');
        dispatch(editExpense(id, updates));
      })
      .catch((e) => {
        //  console.log('Err : ', e);
      });
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
        // console.log(snapshot.val());
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        // console.log(expenses);
        dispatch(setExpenses(expenses));
      })
      .catch((e) => {
        // console.log('error fetching data', e);
      });
  };
};
