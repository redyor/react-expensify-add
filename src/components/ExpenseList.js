import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';
export const ExpenseList = (props) => (
  <div>
    <h3>Expense List</h3>
    {props.expenses.length === 0 ? (
      <p>No Expenses</p>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectedExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
