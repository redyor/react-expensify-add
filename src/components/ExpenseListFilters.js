import React from 'react';
import 'react-dates/initialize';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
    console.log(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'amount') {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              type="text"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDateId="MyDatePicker"
              endDateId="MyEndDateId"
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
