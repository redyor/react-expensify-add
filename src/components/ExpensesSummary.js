import React from 'react';
import { connect } from 'react-redux';
import selectedExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expenses-total';
import numeral from 'numeral';
export const ExpensesSummuary = (props) => (
  <div>
    <p>
      Viewing {props.expenses.length}{' '}
      {props.expenses.length > 1 ? <span>expenses</span> : <span>expense</span>}{' '}
      totaling {numeral(props.total / 100).format('$0,0.00')}
    </p>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectedExpenses(state.expenses, state.filters),
    total: totalExpenses(state.expenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummuary);
