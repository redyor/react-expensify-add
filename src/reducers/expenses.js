// Expense Reducer
// Default State
const expenseReducerDefaultState = [];

const expenseReducer = (state = expenseReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        // console.log(expense, action.id);
        if (expense.id === action.id) {
          //          console.log(expense);
          return {
            ...expense,
            ...action.updates,
          };
        } else {
          return expense;
        }
      });
    //   const expense = state.find(({ id }) => id === action.expense.id);
    //   const { description, note, amount, createdAt } = action.expense;
    //   return {
    //     ...expense,
    //     description,
    //     amount,
    //     note,
    //     createdAt,
    //   };
    default:
      return state;
  }
};

export default expenseReducer;
