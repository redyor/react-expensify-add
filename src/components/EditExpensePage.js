import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  // constructor(props) {
  //   super(props);
  // }
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  removeExpense = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, ownprops) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === ownprops.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => {
      return dispatch(editExpense(id, expense));
    },
    removeExpense: (id) => {
      return dispatch(removeExpense(id));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
