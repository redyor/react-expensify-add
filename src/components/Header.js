import React from 'react';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import { NavLink } from 'react-router-dom';
export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify!</h1>
    <NavLink activeClassName="is-active" to="/dashboard">
      Dashboard
    </NavLink>
    <NavLink activeClassName="is-active" to="/create">
      Create Expense
    </NavLink>

    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchtoProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchtoProps)(Header);
