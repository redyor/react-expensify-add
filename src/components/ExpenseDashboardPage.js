import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummuary from './ExpensesSummary';
const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummuary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default ExpenseDashboardPage;
