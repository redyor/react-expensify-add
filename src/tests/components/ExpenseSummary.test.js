import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummuary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import getExpensesTotal from '../../selectors/expenses-total';

it('should render properly all expenses more than 1', () => {
  const total = getExpensesTotal(expenses);
  const wrapper = shallow(
    <ExpensesSummuary expenses={expenses} total={total} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should render properly 1 expense more than 1', () => {
  const total = getExpensesTotal([expenses[1]]);
  const wrapper = shallow(
    <ExpensesSummuary expenses={[expenses[1]]} total={total} />
  );
  expect(wrapper).toMatchSnapshot();
});
