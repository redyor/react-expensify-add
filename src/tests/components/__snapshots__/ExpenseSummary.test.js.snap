// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1 expense more than 1 1`] = `
<div>
  <p>
    Viewing 
    1
     
    <span>
      expense
    </span>
     
    totaling 
    $14.00
  </p>
</div>
`;

exports[`should render properly all expenses more than 1 1`] = `
<div>
  <p>
    Viewing 
    3
     
    <span>
      expenses
    </span>
     
    totaling 
    $63.00
  </p>
</div>
`;
